### PR TITLE
Unit test proposal

### PR DESCRIPTION
**Pull request by marcoala**
https://github.com/festicket/front-end-meta/blob/FES-7546-unit-test/testing/unit-test.md

### CODE REVIEW

File: testing/unit-test.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  Changing the code to fit the test may sound counterintuitive, but if a piece of code is easy to test that means that is very modular and so very maintainable.
+
+- Write big set of fixtures for a test, same reason as above, a function should only take a very limited set of parameters. If you need too much data to test something that is sign that it's time to break down your code.
+
+## TDD
+Unit test are a very good to practise Test Driven Development, it's not the purpose of this guide to cover the exhaustively the subject, but just to provide a quick to consult cheat sheet for TDD.
+
+1) Write Function definition with empty implementation.  
+2) Write a small doc for the function, you should be able to explain every function in a sentence.  
+3) Write all the test that you need for every function, they should all fail now, skip them and commit.
```
Comment (cutandpastey) [inline testing/unit-test.md]: I fundamentally disagree with this i'm afraid. My workflow works like this:

- Write a test for a small piece of the overall functionality.
- Make it pass.
- Refactor if required.
- Move onto the next test.

99% of the time i work on something I dont know what its going to do completely. Writing every test upfront would slow me down significantly. Its fine to go back and delete tests if the implementation evolves.

File: testing/unit-test.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Unit test
+
+## Introduction
+Unit test are used to test individual part of the application in isolation. They cover a huge variety of cases so this guideline are very generic.
+
+## Dos
+- Test the API of modules, if something is used outside his context should be robustly tested.
+- Test as documentation: looking at a good set of unit test you should be able to understand how to use a function without reading its internal code.
+- Test edge cases, what happen with undefined parameters, or with empty parameters?
+- Test errors, if a function is expecting a specific type of parameter test that the function is correctly raising errors.
```
Comment (richiemccoll) [inline testing/unit-test.md]: Also if you have a bug in production that you create a fix for, there should be a corresponding unit test.
